File: Easy/1394_Find_Lucky_Integer_in_an_Array/Solution.py
from typing import List, Optional


class Solution:
    def findLucky(self, arr: List[int]) -> int:
        counts = {}

        for num in arr:
            counts[num] = counts.get(num, 0) + 1
        
        lucky_integer = -1

        for num, freq in counts.items():
            if num == freq:
                lucky_integer = max(lucky_integer, num)
                
        return lucky_integer
        


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    arr1 = [2, 2, 3, 4]
    print(solution.findLucky(arr1))  # Output: 2
    
    # test cases 2
    arr2 = [1, 2, 2, 3, 3, 3]
    print(solution.findLucky(arr2))  # Output: 3

    # test cases 3
    arr3 = [2, 2, 2, 3, 3]
    print(solution.findLucky(arr3))  # Output: -1
    
    
    

File: Easy/1394_Find_Lucky_Integer_in_an_Array/Solution.java


public class Solution {
    public int findLucky(int[] arr) {
        int[] count = new int[501];

        for (int num : arr) {
            count[num]++;
        }

        int luckyInteger = -1;

        for (int i = 1; i < count.length; i++) {
            if (count[i] == i) {
                luckyInteger = i;
            }
        }

        return luckyInteger;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        int[] arr1 = {2, 2, 3, 4};
        System.out.println(solution.findLucky(arr1)); // Output: 2
        
        // test cases 2
        int[] arr2 = {1, 2, 2, 3, 3, 3};
        System.out.println(solution.findLucky(arr2)); // Output: 3

        // test cases 3
        int[] arr3 = {2, 2, 2, 3, 3};
        System.out.println(solution.findLucky(arr3)); // Output: -1
        
    }
}